eos-metapackage

```diff
@@ -71,16 +81,52 @@ sub get_depends {
 	return @deps;
 }
 
+sub check_installable {
+	my ($depfile, $deps) = @_;
+	my @cmd;
+	my %options = (
+		# ignore packages for foreign architectures; they may not be
+		# installable on this architecture, and that's OK.
+		reduce_restrictions => 1,
+	);
+
+	verbose_print("checking all packages in $depfile are installable");
+	deps_iterate(deps_parse($deps, %options), sub {
+		my ($d) = @_;
+		my $package = $d->{package};
+
+		if (grep { $package eq $_ } @{$dh{DOPACKAGES}}) {
+			verbose_print("skipping '$package' which comes from this project");
+		} else {
+			push @cmd, $package;
+		}
+
+		return 1; # keep iteratin'
+	});
+
+	if (@cmd > 0) {
+		unshift(@cmd, ("apt-get", "install", "--download-only", "--dry-run", "--quiet"));
```

Comment:
This just doesn't work at all when building in my OBS branch. https://obs-master.endlessm-sf.com/build/home:wjt:branches:eos:master:endless/endless/x86_64/eos-meta/_log#

```
[  118s] DH_VERBOSE='1' /usr/src/packages/BUILD/eos-metapackage --check-params
[  118s] --- Listing only the last 100 lines from a long log. ---
[  118s] E: Unable to locate package gedit
[  118s] E: Unable to locate package gkbd-capplet
[  118s] E: Unable to locate package gnome-calculator
[  118s] E: Unable to locate package gnome-clocks
[  118s] E: Unable to locate package gnome-contacts
[  118s] E: Unable to locate package gnome-control-center
...
```

Presumably because there's no apt cache in the build chroot (and there's no network so I can't just chuck an `apt-get update` in there). So… I want the build to behave differently under `jenkins-obs` and `obs`, and I think the easiest way to achieve this is a build option that is enabled by default and disabled in `debian/rules`.

Reply:
Yeah, it will be difficult to make this work in OBS. There's a repo at `/.build.binaries`, but it's constructed per build and only includes the packages OBS determines are needed for the build. It's not a full repo and certainly not a representation of our shared repo.

Reply:
I “solved” this by adding an on-by-default option in `meson_options.txt`, and disabling it in `debian/rules`. This does of course mean that local builds with `dpkg-buildpackage -b` won't check this, but the goal is to check PRs, really.

Reply:
That works for now. I was thinking about this a little more, though. One issue with only running it in jenkins is that it will only check amd64 and not armhf. It would be ideal to have this run in OBS on all architectures.

How could you run this in OBS? Setup a temporary apt configuration like `chdist` does (or we've done 8 million times with `python-apt`). If you look at `dist_create` in `chdist`, you can see how it does this. Then you just run `apt` overriding the `APT_CONFIG` environment variable (or you could pass the `--config-file` option). One advantage of this approach is it points to an empty dpkg status file, so apt isn't influenced by the packages installed in the buildroot. You could do this yourself by making an empty temporary file and then setting the apt option `Dir::State::status` to it.

However, you'd still be a little stuck because you wouldn't know what dist to put in the sources file. Ideally you'd use something from `/etc/os-release` or `/etc/lsb-release` (assuming you have the proper build dep to pull them in). However, we don't actually fill those out with our codename correctly. We don't have `VERSION_CODENAME` in `/etc/os-release` and `DISTRIB_CODENAME` in `/etc/lsb-release` is just set to a generic `endless`. We'd want one of those to be `master` or `eosX.Y`. That would require a little rework of our branching process, though. Likewise, you don't know the components, but I think you could just hardcode `core endless` in there.

Anyways, food for thought but I think we can just go ahead with what you have as it's already a major improvement.